### PR TITLE
NLL Update for Tranquilizer Rifle

### PR DESCRIPTION
**Pull request by Gamersxp6248**
NLL can be included by any weapon swep, and NLL.PlayerZZZ() can be called on hit to make the gun work as expected. Currently uses a method of shooting that may ignore clientside hitreg. May be fixed later.

### CODE REVIEW

File: lua/weapons/weapon_ttt_m16_stn.lua
Comment (TW1STaL1CKY) [inline lua/weapons/weapon_ttt_m16_stn.lua:37]: This field tells TTT it should spawn on the map by default at the specified spawn node type. If this is meant to only be available in the shop (which I assume it is), this should be left out.

File: lua/weapons/weapon_ttt_m16_stn.lua
Comment (TW1STaL1CKY) [inline lua/weapons/weapon_ttt_m16_stn.lua:36]: Don't set this - only built-in TTT weapons should have this field

File: lua/weapons/weapon_ttt_m16_stn.lua
Comment (TW1STaL1CKY) [inline lua/weapons/weapon_ttt_m16_stn.lua:9]: By the way, if you want this to work **exactly** like the M16 just with a tweaks to how it fires, you can use the M16 SWEP as the baseclass. Then you don't need to copy/rewrite everything from the M16, you can just overwrite what you need to.

File: lua/weapons/weapon_ttt_rifle_stn.lua
Comment (TW1STaL1CKY) [inline lua/weapons/weapon_ttt_rifle_stn.lua:9]: Same as what I wrote about the M16 - if you want this to work exactly like the Rifle but with a few tweaks, you can use the Rifle SWEP as the baseclass. Then you don't need to copy/rewrite everything from the Rifle, you can just overwrite what you need to.

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:38]: What stops you from using `EntityTakeDamage` or similar to find out if a player has taken damage instead of iterating through each player each tick?

File: lua/weapons/weapon_ttt_rifle_stn.lua
Comment (TW1STaL1CKY) [inline lua/weapons/weapon_ttt_rifle_stn.lua:37]: Same as what I mentioned on the M16 - remove `SWEP.builtin` and reconsider `SWEP.spawnType`

File: lua/weapons/weapon_ttt_rifle_zzz.lua
Comment (TW1STaL1CKY) [inline lua/weapons/weapon_ttt_rifle_zzz.lua:9]: Ditto comments for this weapon

File: lua/autorun/nl_library.lua
Comment (jule14) [inline lua/autorun/nl_library.lua:12]: dumb the networking down to just one networkstring

**Review by jule14 — CHANGES_REQUESTED** [review]

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:439]: Use a Convar with userinfo enabled and `ply:GetInfoNum()` for this instead of concommand. You don't need to manually network it and can get rid of the net message entirely.

**Review by Zeni44 — CHANGES_REQUESTED** [review]
These are some parts I'd like to see get looked at before it gets merged.
That aside, overall `//` and `!=` isn't supported by Githubs highlighting in lua files but also isn't supported outside of Gmod. While it does come down to personal preference since this is Gmod, it's usually better practice not to use it.

File: lua/autorun/nl_library.lua
Comment (Lixquid) [inline lua/autorun/nl_library.lua:1]: Prefer pure lua syntax over GLua specific addons like `//`

File: lua/autorun/nl_library.lua
Comment (Lixquid) [inline lua/autorun/nl_library.lua:48]: Is this from the tazer addon or a painfully generic field that wasn't polish notation namespaced

File: lua/weapons/weapon_ttt_m16_stn.lua
Comment (TW1STaL1CKY) [reply lua/weapons/weapon_ttt_m16_stn.lua:37]: When setting the base to M16, set this field to nil instead

File: lua/autorun/nl_library.lua
Comment (Lixquid) [inline lua/autorun/nl_library.lua:18]: if possible move this data attachment stuff to a table in your library instead of attaching onto players directly

File: lua/weapons/weapon_ttt_m16_stn.lua
Comment (Lixquid) [inline lua/weapons/weapon_ttt_m16_stn.lua:117]: Formatting

File: lua/weapons/weapon_ttt_rifle_stn.lua
Comment (Lixquid) [inline lua/weapons/weapon_ttt_rifle_stn.lua:105]: formatting

File: lua/weapons/weapon_ttt_rifle_zzz.lua
Comment (Lixquid) [inline lua/weapons/weapon_ttt_rifle_zzz.lua:105]: formatting

File: lua/autorun/nl_library.lua
Comment (Lixquid) [inline lua/autorun/nl_library.lua:228]: Condense this to a single network call

**Review by Lixquid — CHANGES_REQUESTED** [review]

**Comment by Earu** [comment]
Considering the amount of work this PR has received already, and the volume of the changes for just a weapon, I am concerned whether this adds any real value other than the existing similar weapons (stungun?)

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:2]: After looking through each file, this does not seem to be referenced outside of this file.
This also does not need to be a global, slipped through the first time.

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:34]: Why is the second return a string? Better yet, why is there a second return at all?

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:215]: `ply.condition` and `NLL.ConditionStorage` achieve the same thing yet it's stored separately twice and used differently to achieve the same result. Get rid of one of them and make the rest of the code use the same method.

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:295]: Ditto

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:459]: Same deal with using two different existing systems. Decide on either `ply.condition` or `NLL.ConditionStorage`.

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:379]: Unclean, `NLLSleepIn` uses the same code twice. This should be a localized function to avoid redundancy.

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:397]: Ditto

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:401]: Same deal as above except with `NLLUnragdoll`

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:562]: Ditto.

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:357]: This doesn't check for player validity and can error on player disconnect.

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:411]: Ditto

File: lua/autorun/nl_library.lua
Comment (Zeni44) [inline lua/autorun/nl_library.lua:420]: Ditto

**Review by Zeni44 — CHANGES_REQUESTED** [review]
I really feel like this whole PR needs a rework. This is only the library and I haven't even been checking the rest. Either way this is all I have to suggest for now.

File: lua/autorun/nl_library.lua
Comment (Gamersxp6248) [reply lua/autorun/nl_library.lua:459]: This is the clientside code, which does not need to loop through each conditioned player. Since I am not going to be using a table, I won't be defining a table.

File: lua/autorun/nl_library.lua
Comment (Gamersxp6248) [reply lua/autorun/nl_library.lua:562]: This is the clientside code, which does not need to loop through each conditioned player. Since I am not going to be using a table, I won't be defining a table.

**Comment by Earu** [comment]
> Considering the amount of work this PR has received already, and the volume of the changes for just a weapon, I am concerned whether this adds any real value other than the existing similar weapons (stungun?)
> 
> 
> 
> Maybe efforts are better put somewhere else?

@TW1STaL1CKY @Lixquid @Zeni44 thoughts ?

File: lua/autorun/nl_library.lua
Comment (Zeni44) [reply lua/autorun/nl_library.lua:459]: Then make the table shared? If anything, polluting the player table is worse than having your own stored which is already setup and easy to change.
Any reasonable maintainer would expect the same vars to be stored in the same table clientside and not elsewhere with a different system to achieve the same thing.
This also renders `NLL.GetCondtion` completely pointless as it's unused and never will be used by this logic.

File: lua/autorun/nl_library.lua
Comment (Zeni44) [reply lua/autorun/nl_library.lua:562]: Ditto as other comment.

**Comment by Gamersxp6248** [comment]
If you're only criticism is that it 'requires work' then I'm alarmed that you've created as much as you have in the past. Please highlight issues with the code, not issues with your dedication to fun.

File: lua/autorun/nl_library.lua
Comment (Zeni44) [reply lua/autorun/nl_library.lua:430]: Ultimately it doesn't matter on the grand scheme of things, but it is unclean/bad and will make maintaining more bothersome so I will keep it resolved for now.

**Comment by Earu** [comment]
Completely unrelated but that's fine. It's sad to see you've taken this personally. I was only highlighting the fact that there is a lot of dedication made for something which I'm not sure will contribute a whole lot to players experiences.

If you can't understand that maybe you're not fit for contributing here ?

**Review by Earu — CHANGES_REQUESTED** [review]
- Inconsistent casing:
> sometimesnone
> SometimesItsLikeThis

- Please add spaces after commas
- What is NLL ? Make it a proper name so we know what it is
- Make every function you create local by default unless it NEEDS to be global
- Why do you constantly change the attributes set in SWEP.Primary, set a hard value from the start
- More to come...

**Comment by Gamersxp6248** [comment]
> * Inconsistent casing:

Readability ~= Functionality. The pattern that arises is that ply.variable remains uncased, but PlyAssignVariable() is cased.

> * Please add spaces after commas

Readability ~= Functionality. I have zero issue in reading my commas, I'll space them for you in my next commit anyways.

> * What is NLL ? Make it a proper name so we know what it is

Non Lethal Library, not necessary to spell out NonLethalLibrary every time I want to pull a function from the table I made, so it's abbreviated to NLL in the code, would a comment on line 8 detailing -- "Non Lethal Library" suffice?

> * Make every function you create local by default unless it NEEDS to be global

Reasonable advice, I will do this.

> * Why do you constantly change the attributes set in SWEP.Primary, set a hard value from the start

Read comment.

> * Do we actually need 3 weapons that do the same thing? It'd be much less of an issue if it was only one weapon that can be tweaked AND self-contained (no library)

These weapons do not all do the same thing, just because the end result is that 'a player is ragdolled'. Tranquilizer will always knock out on hit, the difference being the time it takes, and it is only usable on the rifle. Players may opt for an alternative style of play, using the stun rifle instead, which is instant, but shorter knock out time. The m16 is offered as a stun variant more or less for the fact it can prevent a trigger happy detective from unwantedly killing an innocent in crossfire.

finally, ragdoll = funny.

> * More to come...

ok

**Comment by Earu** [comment]
These are not up for debate, they are requested changes.

> Non Lethal Library, not necessary to spell out NonLethalLibrary every time I want to pull a function from the table I made, so it's abbreviated to NLL in the code, would a comment on line 8 detailing -- "Non Lethal Library" suffice?

This is only used by the "tranq gun" so just TranqGun should be fine if you **really** need me to pick something up for you.

> These weapons do not all do the same thing, just because the end result is that 'a player is ragdolled'. Tranquilizer will always knock out on hit, the difference being the time it takes, and it is only usable on the rifle. Players may opt for an alternative style of play, using the stun rifle instead, which is instant, but shorter knock out time. The m16 is offered as a stun variant more or less for the fact it can prevent a trigger happy detective from unwantedly killing an innocent in crossfire.

So? Make a first one and base the others on top of it instead of making 3. It also nicely comes back to my point on making the thing self-contained.

**Pull request closed without merging**
